### PR TITLE
Make Background DU configurable to run periodically in order to update used space info on db

### PR DESCRIPTION
**Pull request by enricovianello**
A new Disk Usage Service has been added. It enable admins to run a periodic du on the storage spaces that are not GPFS-with-quota-enabled. By default it's disabled.
It also replaces the previous implementation of the initial du used to initialise the not-initialised storage spaces, that was already present.

Configured into storm.properties with:

```
# Enable/Disable the periodic run of the service
storm.service.du.enabled=false
# Enable/Disable parallel du execution
storm.service.du.parallelTasks=false
# Initial delay of service start in seconds. Default: 60s
storm.service.du.delay=60
# Interval of execution of the du in seconds. Default: 86400s (24h)
storm.service.du.interval=86400
```

### CODE REVIEW

File: etc/storm.properties.template [hunk 1]
```diff
@@ -282,3 +282,20 @@ ptg.skip-acl-setup = false
...
+#
+# Disk Usage service is used for the periodic update of the used-space of all the storage spaces that are not 
+# GPFS-with-quota-enabled. Running a periodic 'du -s -b' on the top of the storage spaces root directory, the
+# used-space stored into the database is updated. By default, the service is disabled.
+# The initial starting delay of the service and the period can also be configured.
+#
+# Enable/Disable the periodic run of the service
+storm.service.du.enabled=false
+# Initial delay of service start in seconds. Default: 60s
+storm.service.du.delay=60
```
Comment (andreaceccanti) [inline etc/storm.properties.template:298]: this should be called delaySecs

File: etc/storm.properties.template [hunk 2]
```diff
@@ -282,3 +282,20 @@ ptg.skip-acl-setup = false
...
+# Interval of execution of the du in seconds. Default: 86400s (24h)
+storm.service.du.interval=86400
```
Comment (andreaceccanti) [inline etc/storm.properties.template:300]: this should be called periodSecs

File: pom.xml [hunk 3]
```diff
@@ -499,6 +499,13 @@
       <scope>test</scope>
     </dependency>
 
+    <!-- https://mvnrepository.com/artifact/mysql/mysql-connector-java -->
```
Comment (andreaceccanti) [inline pom.xml:502]: this is not really needed

File: src/main/java/it/grid/storm/Main.java [hunk 4]
```diff
@@ -15,6 +17,13 @@ public static void main(String[] args) {
 
     StoRM storm = new StoRM();
 
+    try {
+      storm.init();
+    } catch (BootstrapException e) {
+      log.error(e.getMessage(), e);
+      return;
```
Comment (andreaceccanti) [inline src/main/java/it/grid/storm/Main.java:24]: Shouldn't it exit with a non-zero exit code here?

File: src/main/java/it/grid/storm/StoRM.java [hunk 5]
```diff
@@ -444,6 +436,61 @@ public synchronized boolean isExpiredAgentRunning() {
     return isExpiredAgentRunning;
   }
 
+  private void configureDiskUsageService() {
+
+    isDiskUsageServiceEnabled = config.getDiskUsageServiceEnabled();
+
+    NamespaceInterface namespace = NamespaceDirector.getNamespace();
+    List<VirtualFSInterface> toExclude = namespace.getVFSWithQuotaEnabled();
```
Comment (andreaceccanti) [inline src/main/java/it/grid/storm/StoRM.java:444]: toExclude -> quotaEnabledVfs

File: src/main/java/it/grid/storm/StoRM.java [hunk 6]
```diff
@@ -444,6 +436,61 @@ public synchronized boolean isExpiredAgentRunning() {
...
+  /**
+   * Starts the internal timer needed to periodically compute the disk usage for each storage area
+   * configured and update the relative value into storm database.
+   */
+  public synchronized void startDiskUsageService() {
+
+    if (isDiskUsageServiceEnabled) {
+
+      long delay = config.getDiskUsageServiceInitialDelay();
+      long period = config.getDiskUsageServiceTasksInterval();
```
Comment (andreaceccanti) [inline src/main/java/it/grid/storm/StoRM.java:462]: See! it shoud be called periodSecs

File: src/main/java/it/grid/storm/StoRM.java [hunk 7]
```diff
@@ -444,6 +436,61 @@ public synchronized boolean isExpiredAgentRunning() {
...
+
+  /**
+   * Starts the internal timer needed to periodically compute the disk usage for each storage area
+   * configured and update the relative value into storm database.
+   */
+  public synchronized void startDiskUsageService() {
+
+    if (isDiskUsageServiceEnabled) {
+
+      long delay = config.getDiskUsageServiceInitialDelay();
```
Comment (andreaceccanti) [inline src/main/java/it/grid/storm/StoRM.java:461]: it should be called delaySecs

File: src/main/java/it/grid/storm/check/SimpleCheckManager.java [hunk 8]
```diff
@@ -11,20 +11,20 @@
...
-import it.grid.storm.namespace.NamespaceException;
 import it.grid.storm.namespace.VirtualFSInterface;
-import java.util.ArrayList;
-import java.util.Collection;
-import java.util.List;
-import java.util.Map;
-import java.util.Map.Entry;
-import org.slf4j.Logger;
-import org.slf4j.LoggerFactory;
+import jersey.repackaged.com.google.common.collect.Lists;
```
Comment (andreaceccanti) [inline src/main/java/it/grid/storm/check/SimpleCheckManager.java:27]: We import guava, so no need to take this form the jersey repackaged library

File: src/main/java/it/grid/storm/check/SimpleCheckManager.java [hunk 9]
```diff
@@ -63,28 +63,20 @@ protected void loadChecks() {
...
-		try {
-			vfsSet = NamespaceDirector.getNamespace().getAllDefinedVFS();
-		} catch (NamespaceException e) {
-			// never thrown
-			log.error("Unexpected NamespaceException during vfsSet retriving "
-				+ e.getMessage() + " . Unable to add NamespaceFSAssociationCheck");
-			throw new IllegalStateException(
-				"Unexpected NamespaceException from getAllDefinedVFS");
+		if (log.isDebugEnabled()) {
+		  log.debug("Retrieved MountPoints: {}", printMapCoupples(mountPoints));
```
Comment (andreaceccanti) [inline src/main/java/it/grid/storm/check/SimpleCheckManager.java:77]: coupples? The naming can be fixed

File: src/main/java/it/grid/storm/config/Configuration.java [hunk 10]
```diff
@@ -191,6 +191,10 @@
   private static final String NETWORKADDRESS_CACHE_NEGATIVE_TTL =
       "networkaddress.cache.negative.ttl";
 
+  private static final String DISKUSAGE_SERVICE_ENABLED = "storm.service.du.enabled";
+  private static final String DISKUSAGE_SERVICE_INITIAL_DELAY = "storm.service.du.delay";
```
Comment (andreaceccanti) [inline src/main/java/it/grid/storm/config/Configuration.java:195]: delaySecs

File: src/main/java/it/grid/storm/config/Configuration.java [hunk 11]
```diff
@@ -191,6 +191,10 @@
...
+  private static final String DISKUSAGE_SERVICE_TASKS_INTERVAL = "storm.service.du.interval";
```
Comment (andreaceccanti) [inline src/main/java/it/grid/storm/config/Configuration.java:196]: periodSecs

File: src/main/java/it/grid/storm/info/du/DiskUsageExecCommand.java [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+
+import com.google.common.collect.Lists;
+
+import it.grid.storm.space.DUResult;
+
+public class DiskUsageExecCommand {
+
+  private static final Logger log = LoggerFactory.getLogger(DiskUsageExecCommand.class);
+
+  private String absPath;
```
Comment (andreaceccanti) [inline src/main/java/it/grid/storm/info/du/DiskUsageExecCommand.java:22]: should this be final?

File: src/main/java/it/grid/storm/info/du/DiskUsageExecCommand.java [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.collect.Lists;
+
+import it.grid.storm.space.DUResult;
+
+public class DiskUsageExecCommand {
+
+  private static final Logger log = LoggerFactory.getLogger(DiskUsageExecCommand.class);
```
Comment (andreaceccanti) [inline src/main/java/it/grid/storm/info/du/DiskUsageExecCommand.java:20]: log -> LOG

File: src/main/java/it/grid/storm/info/du/DiskUsageService.java [hunk 14]
```diff
@@ -0,0 +1,95 @@
...
+public class DiskUsageService {
+
+  private static final Logger log = LoggerFactory.getLogger(DiskUsageService.class);
+
+  private List<VirtualFSInterface> monitoredSAs;
+
+  private ScheduledExecutorService executor;
+  private boolean running;
+
+  public DiskUsageService(List<VirtualFSInterface> vfss) {
```
Comment (andreaceccanti) [inline src/main/java/it/grid/storm/info/du/DiskUsageService.java:25]: The executor should be a dependency for this service, i.e. passed in as a parameter

File: src/main/java/it/grid/storm/info/du/DiskUsageService.java [hunk 15]
```diff
@@ -0,0 +1,95 @@
...
+
+    return monitoredSAs;
+  }
+
+  public void addMonitoredSA(VirtualFSInterface vfs) {
+
+    monitoredSAs.add(vfs);
+  }
+
+  public int runScheduled(long delay, long period) {
```
Comment (andreaceccanti) [inline src/main/java/it/grid/storm/info/du/DiskUsageService.java:46]: this method should likely be synchronized?

File: src/main/java/it/grid/storm/info/du/DiskUsageService.java [hunk 16]
```diff
@@ -0,0 +1,95 @@
...
+    executor.shutdown();
+    running = false;
+  }
+
+  public boolean isRunning() {
+
+    return running;
+  }
+
+  public static int runTasksOnce(List<VirtualFSInterface> vfss) {
```
Comment (andreaceccanti) [inline src/main/java/it/grid/storm/info/du/DiskUsageService.java:81]: Why is this needed?

File: src/main/java/it/grid/storm/info/du/DiskUsageUtils.java [hunk 17]
```diff
@@ -0,0 +1,71 @@
...
+import java.util.List;
+
+import it.grid.storm.space.DUResult;
+import it.grid.storm.space.ExitCode;
+
+public class DiskUsageUtils {
+
+  private DiskUsageUtils() {}
+
+  private static DUResult getResult(ExitCode code, String absPath, long size, long start, long end) {
```
Comment (andreaceccanti) [inline src/main/java/it/grid/storm/info/du/DiskUsageUtils.java:18]: It would be safer to use new classes from java.time to handle intervals here...

File: etc/storm.properties.template (hunk 1 above)
Comment (enricovianello) [reply etc/storm.properties.template]: fixed

File: pom.xml (hunk 3 above)
Comment (enricovianello) [reply pom.xml]: removed

File: etc/storm.properties.template (hunk 2 above)
Comment (enricovianello) [reply etc/storm.properties.template]: fixed

File: src/main/java/it/grid/storm/Main.java (hunk 4 above)
Comment (enricovianello) [reply src/main/java/it/grid/storm/Main.java]: Fixed with exit(1)

File: src/main/java/it/grid/storm/StoRM.java (hunk 5 above)
Comment (enricovianello) [reply src/main/java/it/grid/storm/StoRM.java]: renamed

File: src/main/java/it/grid/storm/info/du/DiskUsageExecCommand.java (hunk 13 above)
Comment (enricovianello) [reply src/main/java/it/grid/storm/info/du/DiskUsageExecCommand.java]: renamed

File: src/main/java/it/grid/storm/info/du/DiskUsageService.java (hunk 14 above)
Comment (enricovianello) [reply src/main/java/it/grid/storm/info/du/DiskUsageService.java]: The executor is now a parameter with also the list of VFSs, the delay and the period

File: src/main/java/it/grid/storm/info/du/DiskUsageService.java (hunk 15 above)
Comment (enricovianello) [reply src/main/java/it/grid/storm/info/du/DiskUsageService.java]: fixed

File: src/main/java/it/grid/storm/info/du/DiskUsageService.java (hunk 16 above)
Comment (enricovianello) [reply src/main/java/it/grid/storm/info/du/DiskUsageService.java]: Removed and added WARN in the log to notify admins that the used space of one or more storage areas has not been initialised.

File: src/main/java/it/grid/storm/info/du/DiskUsageUtils.java (hunk 17 above)
Comment (enricovianello) [reply src/main/java/it/grid/storm/info/du/DiskUsageUtils.java]: done
